ドラフト

NaoyaFukuma opened the pull request.

NaoyaFukuma committed:
ドラフト

tomofuji-dev reviewed (approved):
確認しました、抜けもれないと思います

NaoyaFukuma committed:
listenディレクティブにおいて、ドメイン名への対応を前提にした修正。

NaoyaFukuma committed:
formatter で改行されていたところを修正

tomofuji-dev reviewed (approved):
OKです！

NaoyaFukuma committed:
.vscodeの混入を修正

NaoyaFukuma committed:
listenディレクティブのデータの持ち方を修正。bind()に備えIPアドレスとポートの重複をなくした対応関係を表現するmapの修正

NaoyaFukuma committed:
config.cppの追加

NaoyaFukuma committed:
parser実装注

NaoyaFukuma committed:
修正中

NaoyaFukuma committed:
修正中

NaoyaFukuma committed:
Merge branch 'main' into 18-config-parser-ステップの出力をチーム内でfixする

NaoyaFukuma committed:
Merge branch 'main' into 18-config-parser-ステップの出力をチーム内でfixする

NaoyaFukuma committed:
コンフィグパーサーの実装

NaoyaFukuma committed:
config.txtは不要。configファイル用のディレクトリ内に、適切なconfファイルを配置

tomofuji-dev committed:
大量リクエスト送付シナリオを追加

tomofuji-dev committed:
引数の明示的な定義を追加

tomofuji-dev committed:
複数クライアントシナリオの追加

Chanmachan commented on docs/conf.md:
```diff
@@ -179,7 +182,7 @@ location /images/,  root /data
-注）複数指定された場合、エラー検出できず、最後に現れた項目の値が採用されます。
-指定が無い場合は、”404 Not Found”か、リソースがヒットするならそのリソースが返されます。
+注）複数指定された場合、エラー検出できず、最初に現れた項目の値が採用されます。
+以下のパターンを許容します。
```
以下が抜けてたりします？

Chanmachan commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,728 @@
+  if (*it_ != c) {
+    // エラー箇所の行数と列数を計算する
+    int row = 1;
+    int column = 1;
```
1の代入はいらなそう

Chanmachan reviewed (approved):
問題なさそうです。テストしながら何かあればリファクタしていく形だと思います。

NaoyaFukuma committed:
conf.md追記。URLのValidation追記。

NaoyaFukuma committed:
エラー箇所検出関数に行数を数えるコードが抜けていたので追記

tomofuji-dev committed:
defaultのclient数を変更

NaoyaFukuma committed:
int vs size_tの部分を修正

NaoyaFukuma committed:
Merge remote-tracking branch 'origin/28-echo-server-テストケースの追加' into 18-config-parser-ステップの出力をチーム内でfixする

NaoyaFukuma committed:
”28-echo-server-テストケースの追加”をmergeした後に、コンパイルできるように修正

NaoyaFukuma committed:
コンパイルエラーのためヘッダーファイルのパスを追加

NaoyaFukuma committed:
Config.hppをインクルードしている部分のconfig.hppに修正。
なぜか、リモートリポジトリ上のファイル名がconfig.hppのままなので、コンパイルエラーになっているため

NaoyaFukuma committed:
Update config.cpp

NaoyaFukuma committed:
リモートリポジトリ上のファイル名変更
”git mv old-filename new-filename”を利用し、
config.hpp -> Config.hpp
config.cpp -> Config.cpp
とする変更を行い。インクルードしている部分を、Config.hppに変更した。

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+      std::string line;
+      this->GetErrorPoint(row, col, line);
+      throw ParserException(ERR_MSG_ROW_COL_LINE, row, col, line.c_str(),
+                            "Expected 'server'");
```
関数にまとめていいと思います

tomofuji-dev replied:
assert内のthrowでもこれを使っていいかなと思いました

NaoyaFukuma replied:
こんな感じにまとめてみました

```c++
// parse error時に、例外をスローする. 第二引数がtrueの場合は、行、列、行の内容も出力する
void ConfigParser::ThrowParseError(const char *msg, bool add_err_point_flag = false) {
  if (add_err_point_flag) {
    int row;
    int col;
    std::string line;
    this->GetErrorPoint(row, col, line);
    throw ParserException(ERR_MSG_ROW_COL_LINE, row, col, line.c_str(), msg);
  } else {
    throw ParserException(ERR_MSG, msg);
  }
}

// parser
void ConfigParser::Parse(Config &config) {
  while (!this->IsEof()) {
    this->SkipSpaces();
    if (this->GetWord() != "server") {
      this->ThrowParseError("Expected 'server'", true);
    }
    this->ParseServer(config);
    this->SkipSpaces();
  }
  AssertConfig(config);
}
```

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+  while (!this->IsEof() && *it_ != '}') {
+    this->SkipSpaces();
+    std::string token = GetWord();
+    if (token == "match") {
```
swtich caseに直したいです

NaoyaFukuma replied:
string型を使って書くことはできますか？

tomofuji-dev replied:
確かに、stringで使えないですね;;

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+  while (!this->IsEof() && isspace(*it_)) {
+    it_++;
+  }
+}
```
\nを他の空白文字と同一視しているので、以下のようなケースも許容します。
webservでは許容することにしてもいいですが、nginxだとparseerrorになるので検討したいです。
```
listen


 hoge:8000
```
SkipSpaceの引数に改行あるなしを指定できるようにして、ディレクティブを読み込んだ後のSkipSpace()を改行なしで実行するとか？

tomofuji-dev replied:
現状だと、以下も許容します
```
listen hoge:8000
;
```

NaoyaFukuma replied:
SkipSpaces()を修正し、スキップ対象に改行を含めるかを設定できるように修正。
そのうえで、改行をスキップしたくない箇所、
ディレクティブとパラメータの間、パラメータとセミコロンの間でのSkipSpaces()呼び出しを修正

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+
+  size_t colon_pos = listen_str.find(':');
+  if (colon_pos == std::string::npos) {
+    host_str = "0.0.0.0";
```
めっちゃ細かいですが、portと同じくhostもvoid ConfigParser::ParseServer(Config &config)の最初の部分で初期化してもいいかも

NaoyaFukuma replied:
Vserver, Locationこれらの構造体のコンストラクタを設けました！

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+
+void ConfigParser::AssertTimeOut(int &dest_timeout,
+                                 const std::string &timeout_str) {
+  if (!ws_strtoi<int>(&dest_timeout, timeout_str)) {
```
以前見た時、timeoutの単位として's'をつけていたと思うのですが、単位をつけないよう変更しました？

tomofuji-dev replied:
's'つけなくてもいいかなと思ってた部分なので、上記仕様ならこのままで大丈夫だと思います

NaoyaFukuma replied:
's'無しに変更しました。Nginxの仕様を調べているうちに、's' はオプション扱いで、省略も許容することに気づき、冗長さを無くし、複数の設定方法があることによる混乱を防ぐために、's' 無ししか許容しない仕様にしました。

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+    throw ParserException(
+        ERR_MSG, (server_name + " is Invalid too long server name").c_str());
+  }
+  for (std::string::const_iterator it = server_name.begin();
```
servernameはip/domain名の両方を許容しますが、ipアドレスのチェックはIsValidLabel（）をループで実行することに含まれているという認識でいいですか？

NaoyaFukuma replied:
たしかに！
IPかドメイン名かの判定を入れて、どちらもValidationします！

NaoyaFukuma replied:
IPアドレスのフォーマットに適合するならIPアドレスとして扱い、
そうでないなら、ドメイン名として扱うように修正しました！

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+      this->GetErrorPoint(row, col, line);
+      throw ParserException(ERR_MSG_ROW_COL_LINE, row, col, line.c_str(),
+                            "Unexpected token");
+    }
```
自分で書いた部分ですが、switch caseに書き直したいです（コンパイラの最適化）

tomofuji-dev commented on docs/conf.md:
```diff
@@ -54,11 +56,12 @@ webservで用いる設定ファイルについての解説。
 - Multiple: false
 - Syntax: `'listen' ((DOMAIN_NAME | IP_ADDR) ':')? (PORT) END_DIRECTIVE`
 - 概要: 仮想サーバーが待ち受けるドメイン名またはIPアドレスとポートを指定します。
 注）複数指定された場合、エラー検出できず、最後に現れた項目の値が採用されます。
```
細かい日本語ですが、「エラー検出できず」ではなく「エラーとして検出せず」がいいです。
仕様でしっかり定義して、あえてそうしてないだけなんだからね！

tomofuji-dev replied:
他も同様の箇所があるので省略

NaoyaFukuma replied:
注）複数指定された場合、エラーとして扱わず、最後に現れた項目の値を採用します。

と修正。エラーとして扱わないことと、最後を採用することを端的な肯定文で記すことで、この挙動が意図したものであることがわかるようにした。
（端的な肯定文でって、日本語あってるのか？）

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+                                     const std::string &method_str) {
+  method_type src_method;
+
+  if (method_str == "GET") {
+    src_method = GET;
+  } else if (method_str == "POST") {
+    src_method = POST;
+  } else if (method_str == "DELETE") {
+    src_method = DELETE;
+  } else {
+    throw ParserException(ERR_MSG, (method_str + " is Invalid method").c_str());
+  }
```
ここもswitch caseに直したいです

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+  }
+}
+
+bool AssertPath(const std::string &path) {
```
自分用メモ：config parseの部分では使用している文字種のチェックだけ　accessできないパスが指定されていたら、対応するerrorレスポンスを返す

tomofuji-dev replied:
boolを返しているので、IsValidPathでもいいかも

NaoyaFukuma replied:
IsValidPathに修正

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+    throw ParserException(ERR_MSG, (std::string(unit_str) +
+                                    " is Invalid client_max_body_size unit")
+                                       .c_str());
+  }
```
try catchでstd::overflow_errorを検知して、parserExceptionとして再throwしたいです

tomofuji-dev replied:
それか、overflowのチェック関数をis_mul_overflow()とboolを返すように書き直すのでもいいと思います。

NaoyaFukuma replied:
try catchで囲みました！

```c++
  try {
    switch (unit) {
    case 'B':
      break;
    case 'K':
      dest_size = mul_assert_overflow<uint64_t>(dest_size, 1024);
      break;
    case 'M':
      dest_size = mul_assert_overflow<uint64_t>(dest_size, 1024 * 1024);
      break;
    case 'G':
      dest_size = mul_assert_overflow<uint64_t>(dest_size, 1024 * 1024 * 1024);
      break;
    default:
      char unit_str[4] = {'\'', unit, '\'', '\0'};
      throw ParserException(ERR_MSG, (std::string(unit_str) +
                                      " is Invalid client_max_body_size unit")
                                         .c_str());
    }
  } catch (const std::overflow_error &e) {
      throw ParserException(ERR_MSG, (std::string(size_str) +
                                      " is Invalid 'uint64_t' overflow client_max_body_size")
                                         .c_str());
    
  }
```

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+  }
+}
+
+void ConfigParser::AssertIndex(const std::string &index) {
```
// indexディレクティブは、rootディレクティブが設定されている場合のみ有効
-> rootは必須項目だと思います

NaoyaFukuma replied:
コメントを削除

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+        (root + " is Invalid root path. use Invalid character.").c_str());
+  }
+}
+
```
root + indexみたいな形で使うと思うので、最後の文字が'/'であることをことを保証しておくといいと思いました
indexが/で始まる場合でも、そうでない場合でも、root + indexでpathが通るようになると思います

tomofuji-dev replied:
追記：parserの時点でそれを行うのではなく、実行時にroot + '/' + indexの形にするのがいい気がしてきました

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+      location.return_.return_type_ = RETURN_ONLY_STATUS_CODE;
+    } else {
+      tmp_str = GetWord();
+      if (tmp_str[0] == '\'') {
```
最後が'であるチェックもした方がいいと思います

tomofuji-dev replied:
最後が'でなければ、URLとしてみなす

NaoyaFukuma replied:
先頭のみしかチェックしていなかったものを、末尾のチェックも追加対応。

tomofuji-dev replied:
"'"がこの条件を突破するので、size() >= 2 をチェックした方が良さそうです

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+      if (tmp_str[0] == '\'') {
+        // text
+        location.return_.return_type_ = RETURN_TEXT;
+        location.return_.return_text_ = tmp_str.substr(1, tmp_str.size() - 2);
```
tmp_str.size() < 2の時オーバーフローします

NaoyaFukuma replied:
当該処理の直前のif文を書き直したことで、２文字以上が保証されるようになったため、修正不要と判断。

tomofuji-dev replied:
GetWord()の結果が"'"だった場合、if文の条件を突破してsize() - 2でオーバーフローすると思います

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+      } else {
+        // URL
+        location.return_.return_type_ = RETURN_URL;
+        location.return_.return_url_ = tmp_str;
```
ここでurlのassertはしなくていいですか？

NaoyaFukuma replied:
以下のassert関数でURLのチェックを行えるようにしました！
void ConfigParser::AssertReturn()
546行目

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,722 @@
+  } else { // status codeがない場合はURLとみなす
+    location.return_.status_code_ = 302; // defaultはリダレクト
+    location.return_.return_type_ = RETURN_URL;
+    location.return_.return_url_ = tmp_str;
```
同上、urlのassertについて

NaoyaFukuma replied:
同上

tomofuji-dev commented on docs/config/valid_case/full_test.conf:
単体テストの作成にあたって、invalid, validともにtestケースを作りたいです
ここら辺はちゃんまさんの作業と合わせて、3人でやりましょう

tomofuji-dev reviewed (changes requested):
ありがとうございます、ConfigParser.cpp, Config.cpp, conf.mdを中心に確認しました！
いくつかコメントしましたので、よろしくお願いします。

NaoyaFukuma committed:
config 仕様書の修正

NaoyaFukuma committed:
parse error時の処理を関数化
プロトタイプ）
// parse error時に、例外をスローする.
// 第二引数がtrueの場合は、行、列、行の内容も出力する
void configParser::ThrowParseError(const char *msg,
                                   bool add_err_point_flag = false);

NaoyaFukuma committed:
ThrowParseError()への差し替え

NaoyaFukuma committed:
isspace()を自作
使用可能関数にないため

NaoyaFukuma committed:
SkipSpaces()を拡張
スキップする文字を選択できる。
デフォルトは、改行を含む。
オプションをtrueにする（引数で指定）と改行を含まない。
ディレクティブとパラメータの間、
パラメータとセミコロンの間では改行を許容しない。

NaoyaFukuma committed:
コンパイルエラーになっていたので修正

NaoyaFukuma committed:
コンパイルエラーを修正

NaoyaFukuma committed:
SkipSpaces()呼び出しを変更
改行を読み飛ばしたくない箇所である、
ディレクティブとパラメータの間、
パラメータとセミコロンの間で、
SkipSpaces()を呼び出す際に、引数にtrueを設定。

NaoyaFukuma committed:
Vserver, Locationにコンストラクタを追加。
上記２つの構造体において、
デフォルトの設定が存在するディレクティブを対象にパラメータを設定。

NaoyaFukuma committed:
server_nameがIPアドレスだった場合のValidationに対応
IPアドレスのフォーマットに適合するならIPアドレスとして扱い、
そうでないなら、ドメイン名として扱う。

NaoyaFukuma committed:
AsseertPath() -> isValidPath()に修正。
戻り値がboolであり、この関数の役割にAssertは適さず、is...の役割だから

NaoyaFukuma committed:
関数名の表記ゆれを修正

NaoyaFukuma committed:
over flow検知漏れを修正。
mul_assert_overflow()がthrowする例外をcatchしていなかった部分を修正。

NaoyaFukuma committed:
rootディレクティブが必須項目ではないことを前提にした間違ったコメントを削除

NaoyaFukuma committed:
returnディレクティブでURL
かテキストかの判定を修正
末尾のチェックも追加

Chanmachan commented:
status_code_って指定がない場合って初期化とかって必要ないですか？

NaoyaFukuma commented:
@Chanmachan 
> status_code_って指定がない場合って初期化とかって必要ないですか？

return ディレクトリが.confに存在しない場合、
enum return_type が、EMPTYになっているので、
status_code_がHTTPリクエスト処理時に読み込まれることはない。という前提から、
return ディレクトリが存在しない場合は、status_code_が未初期化かつ値の代入がされない状態を許容しています。

266行目の以下の部分のことですか？

  } else { // status codeがない場合はURLとみなす
    location.return_.status_code_ = 302; // defaultはリダレクト
    location.return_.return_type_ = RETURN_URL;
    location.return_.return_url_ = tmp_str;
  }

NaoyaFukuma committed:
location path のValidationを追加。
ファイルシステム上、有効な文字種であるかのチェックのみ。アクセス可能か？は、config段階では行わない。

tomofuji-dev commented on srcs/config/ConfigParser.cpp:
```diff
@@ -0,0 +1,735 @@
+void ConfigParser::ParseServer(Config &config) {
+  Vserver server;
+
+  this->SkipSpaces();
```
true

tomofuji-dev reviewed (approved):
2点だけ追加でコメントしました。mergeお願いします！

NaoyaFukuma committed:
serverディレクティブで、"server"と"{" の間に改行を許容するバグがあったので、修正

NaoyaFukuma committed:
returnディレクティブで指定されたテキストが２文字以上であることを保証する
２文字以上無ければURLとして扱う

NaoyaFukuma committed:
stringクラスのbackメソッドが、C++９８では使用できないため修正
